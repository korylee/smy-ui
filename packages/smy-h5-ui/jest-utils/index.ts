import Vue from 'vue'
import { Wrapper } from '@vue/test-utils'
import { IN_BROWSER } from '@smy-h5/shared'

Vue.config.productionTip = false

export const delay = (time = 0) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window | Document, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function trigger(
  wrapper: Wrapper<Vue> | HTMLElement | Window | Document,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0,
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = document.createEvent('CustomEvent')
  event.initCustomEvent(eventName, true, true, {})

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })
  el.dispatchEvent(event)

  return Vue.nextTick()
}

export async function triggerDrag(el: HTMLElement, relativeX: number, relativeY: number) {
  let x = relativeX
  let y = relativeY
  let startX = 0
  let startY = 0
  if (relativeX < 0) {
    startX = Math.abs(relativeX)
    x = 0
  }
  if (relativeY < 0) {
    startY = Math.abs(relativeY)
    y = 0
  }
  await trigger(el, 'touchstart', startX, startY)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await delay(300)
  await trigger(el, 'touchend', x, y)
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method] as any
  console[method] = fn
  return () => {
    console[method] = originMethod
  }
}

export function mockGetBoundingClientRect(rect: Partial<DOMRect>): () => void {
  if (IN_BROWSER) {
    const spy = jest.spyOn(Element.prototype, 'getBoundingClientRect')
    spy.mockReturnValue(rect as DOMRect)
    return () => spy.mockRestore()
  }
  return () => {}
}
