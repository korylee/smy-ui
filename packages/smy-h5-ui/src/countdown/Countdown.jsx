import { createNamespace } from '../_utils/vue/create'
import { props } from './props'
import { parseTime, useCountdown, formatTime } from './utils'

const [name, bem] = createNamespace('countdown')

export default {
  name,
  props,

  data: (vm) => ({
    timeData: {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      milliseconds: 0,
      total: 0,
    },
    countdown: useCountdown({
      onChange: (time) => {
        const timeData = formatTime(time)
        vm.timeData = timeData
        vm.$emit('change', timeData)
      },
      onStart: (time) => {
        vm.$emit('start', time)
        vm.$emit('update:paused', false)
      },
      onEnd: () => {
        vm.$emit('end')
      },
      onPause: (time) => {
        vm.$emit('pause', time)
        vm.$emit('update:paused', true)
      },
    }),
  }),

  watch: {
    time: { immediate: true, handler: 'init' },
    paused(val, oldVal) {
      const { countdown } = this
      const isStart = countdown.isStart()
      if (!oldVal && isStart) this.pause()
      else if (oldVal && !isStart) this.start()
    },
  },

  beforeDestroy() {
    this.pause()
  },

  methods: {
    init() {
      const { autoStart, paused, start, reset } = this
      const run = autoStart && !paused ? start : reset
      run()
    },
    start() {
      this.countdown.start(this.time)
    },
    pause() {
      this.countdown.pause()
    },
    reset() {
      this.countdown.reset(this.time)
    },
  },
  render() {
    const _vm = this
    const _h = _vm.$createElement
    const _c = _vm._self._c || _h
    const { timeData } = _vm

    return _c(
      'div',
      { staticClass: bem(), role: 'timer' },
      [_vm._t('default', () => [parseTime(_vm.format, timeData)], timeData)],
      2,
    )
  },
}
