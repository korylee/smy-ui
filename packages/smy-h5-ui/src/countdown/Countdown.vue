<template>
  <div class="smy-countdown">
    <slot v-bind="timeData"> {{ parseTime(format, timeData) }}</slot>
  </div>
</template>

<script>
import { createNamespace } from '../_utils/vue/create'
import { props } from './props'
import { parseTime, useCountdown, formatTime } from './utils'

const [name] = createNamespace('countdown')

export default {
  name,
  props,

  data: (vm) => ({
    timeData: {
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      milliseconds: 0,
      total: 0,
    },
    countdown: useCountdown({
      onChange: (time) => {
        const timeData = formatTime(time)
        vm.timeData = timeData
        vm.$emit('change', timeData)
      },
      onStart: (time) => {
        vm.$emit('start', time)
        vm.$emit('update:paused', false)
      },
      onEnd: () => {
        vm.$emit('end')
      },
      onPause: (time) => {
        vm.$emit('pause', time)
        vm.$emit('update:paused', true)
      },
    }),
  }),

  watch: {
    time: 'reset',
    paused(val, oldVal) {
      const { countdown } = this
      const isStart = countdown.isStart()
      if (!oldVal && isStart) this.pause()
      else if (oldVal && !isStart) this.start()
    },
  },

  created() {
    const { autoStart, paused, start, reset } = this
    const run = autoStart && !paused ? start : reset
    run()
  },

  beforeDestroy() {
    this.pause()
  },

  methods: {
    parseTime,
    start() {
      this.countdown.start(this.time)
    },
    pause() {
      this.countdown.pause()
    },
    reset() {
      this.countdown.reset(this.time)
    },
  },
}
</script>
