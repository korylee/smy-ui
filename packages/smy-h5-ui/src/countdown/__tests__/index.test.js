import { mount } from '@vue/test-utils'
import Vue from 'vue'
import Countdown from '..'
import { delay } from '../../../jest-utils'

test('test countdown plugin', () => {
  Vue.use(Countdown)
  expect(Vue.component(Countdown.name)).toBeTruthy()
})

describe('test countdown props', () => {
  test('test format prop', () => {
    const wrapper = mount(Countdown, {
      propsData: {
        time: 108000000,
        format: 'HH-mm-ss-SS',
      },
    })
    const reg = /(\d{2}-){3}\d{2}/
    expect(reg.test(wrapper.text())).toBe(true)
    wrapper.destroy()
  })
  test('test countdown auto-start prop', async () => {
    const wrapper = mount(Countdown, {
      propsData: {
        time: 10800,
        autoStart: false,
      },
    })

    const text = wrapper.text()

    await delay(100)

    expect(wrapper.text()).toBe(text)
    wrapper.destroy()
  })
})

describe('test countdown events', () => {
  test('test countdown onEnd and onChange', async () => {
    const onEnd = jest.fn()
    const onChange = jest.fn()

    const wrapper = mount(Countdown, {
      propsData: {
        time: 1000,
        autoStart: true,
      },
      listeners: {
        end: onEnd,
        change: onChange,
      },
    })

    await delay(1100)

    expect(onEnd).toHaveBeenCalledTimes(1)
    expect(onChange).toHaveBeenCalled()
    wrapper.destroy()
  })

  test('test onChange event argument', async () => {
    const onChange = jest.fn()
    const wrapper = mount(Countdown, {
      listeners: {
        change: onChange,
      },
      propsData: { time: 1, autoStart: true, millisecond: true },
    })
    await delay(50)

    expect(onChange.mock.calls[0][0]).toEqual({
      days: 0,
      hours: 0,
      minutes: 0,
      seconds: 0,
      milliseconds: 1,
      total: 1,
    })

    wrapper.destroy()
  })
})

test('test start, payse ant reset methods', async () => {
  const template = `<smy-countdown :time="time" ref="countdown" :auto-start="false"></smy-countdown>`
  const wrapper = mount({
    template,
    components: { SmyCountdown: Countdown },
    data: () => ({ time: 108000 }),
  })
  const text = wrapper.text()

  await delay(100)
  expect(wrapper.text()).toBe(text)
  wrapper.vm.$refs.countdown.start()
  await delay(100)
  expect(wrapper.text()).not.toBe(text)

  wrapper.vm.$refs.countdown.pause()
  const pauseText = wrapper.text()
  await delay(100)
  expect(wrapper.text()).toBe(pauseText)

  wrapper.vm.$refs.countdown.reset()
  await delay(500)
  expect(wrapper.text()).toBe(text)

  wrapper.destroy()
})
