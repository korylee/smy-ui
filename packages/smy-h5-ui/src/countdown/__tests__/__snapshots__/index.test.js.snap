// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test countdown example 1`] = `
"<div class=\\"smy-countdown-example\\">
  <div class=\\"app-demo-title\\">基础使用</div>
  <div class=\\"smy-countdown\\"> 03 : 00 : 00</div>
  <div class=\\"app-demo-title\\">自定义格式化</div>
  <div class=\\"smy-countdown\\"> 00 天 03 时 00 分 00 秒</div>
  <div class=\\"app-demo-title\\">展示毫秒</div>
  <div class=\\"smy-countdown\\"> 30 : 00 : 00 : 000</div>
  <div class=\\"app-demo-title\\">控制开始和暂停的倒计时</div>
  <div class=\\"smy-cell smy-cell--borderless\\">
    <!---->
    <div class=\\"smy-cell__content\\">
      <div class=\\"smy-cell__title\\">
        <div class=\\"smy-countdown\\"> 03 : 00 : 00</div>
      </div>
      <!---->
    </div>
    <div class=\\"smy-cell__extra\\"><button>start</button></div>
  </div>
  <div class=\\"app-demo-title\\">自定义样式</div>
  <div class=\\"smy-countdown\\"><span class=\\"block\\">6</span> <span class=\\"colon\\">:</span> <span class=\\"block\\">0</span> <span class=\\"colon\\">:</span> <span class=\\"block\\">0</span></div>
  <div class=\\"app-demo-title\\">手动控制</div>
  <div class=\\"smy-countdown\\"> 03 : 000</div>
  <div class=\\"btn-container\\"><button>start</button> <button>pause</button> <button>reset</button></div>
</div>"
`;
